crates/bitfinity-block-validator/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+//! Bitfinity block validator.
+
+use did::unsafe_blocks::ValidateUnsafeBlockArgs;
+use did::{Transaction, H256};
+use evm_canister_client::{CanisterClient, EvmCanisterClient};
+use itertools::Itertools;
+use reth_db::database::Database;
+use reth_engine_tree::tree::MemoryOverlayStateProvider;
+use reth_evm::execute::Executor as _;
+use reth_evm::execute::{BlockExecutionOutput, BlockExecutorProvider as _};
+use reth_evm_ethereum::execute::EthExecutorProvider;
+use reth_evm_ethereum::{execute::EthBlockExecutor, EthEvmConfig};
+use reth_primitives::U256;
+use reth_primitives::{Address, Receipt};
+use reth_primitives::{Block, BlockWithSenders};
+use reth_provider::{ChainSpecProvider as _, ExecutionOutcome, ProviderFactory, StateProvider};
+use reth_revm::database::StateProviderDatabase;
+
+/// Block validator for Bitfinity.
+///
+/// The validator validates the block by executing it and then
+/// confirming it on the EVM.
+#[derive(Clone, Debug)]
+pub struct BitfinityBlockValidator<C, DB>
+where
+    C: CanisterClient,
+{
+    evm_client: EvmCanisterClient<C>,
+    provider_factory: ProviderFactory<DB>,
+}
+
+impl<C, DB> BitfinityBlockValidator<C, DB>
+where
+    C: CanisterClient,
+    DB: Database,
+{
+    /// Create a new [`BitfinityBlockValidator`].
+    pub fn new(evm_client: EvmCanisterClient<C>, provider_factory: ProviderFactory<DB>) -> Self {
+        Self { evm_client, provider_factory }
+    }
+
+    /// Validate a block.
+    pub async fn validate_block(
+        &self,
+        block: Block,
+        transactions: &[Transaction],
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        if !self.unsafe_blocks_enabled().await? {
+            tracing::debug!("Unsafe blocks are disabled");
+            return Ok(());
+        }
```

Comment:
Does this check have any meaning? If unsafe blocks are disabled on EVM wouldn't it just return empty list if we request them? In this case we anyway do one GET request on every iteration. But if it's enabled, with this check on every iteration we do 2 GET requests instead of 1.